Add support for Epinio's Dex Auth Provider

richard-cox opened the pull request:
Contributes to https://github.com/epinio/ui/issues/167, https://github.com/epinio/ui/issues/176

Related PRs https://github.com/rancher/dashboard/pull/7299, https://github.com/epinio/helm-charts/pull/295

richard-cox committed:
Add support for Epinio's Dex Auth Provider

richard-cox commented:
@nwmac This is the backend work for dex, i'm sure there's some golang smells here and possible the way i've integrated dex

- there's still a single epinio auth type that covers both methods of logging in ('local' user and dex). I don't are only model only supported a single way to log in
- the dex provider didn't work with the existing oidc/oauth workflow so there's a new dex request handler that works in the similar way as oidc/oauth request handling

richard-cox committed:
Fix linting

richard-cox committed:
Fix tests

thehejik commented on src/jetstream/plugins/epinio/main.go:
```diff
@@ -36,30 +42,51 @@ func init() {
-	log.Infof("Epinio API url: '%s'. Epinio WSS url: '%s'. Skipping SSL Validation: '%+v'", epinioApiUrlValue, epinioApiWsUrlValue, epinioApiUrlskipSSLValidation)
+	epinioAuthUrlValue, _ := portalProxy.Env().Lookup(epinioDexAuthUrl)
+	if len(epinioAuthUrlValue) == 0 {
+		epinioAuthUrlValue = strings.Replace(epinioApiUrlValue, "epinio.", "auth.", 1)
```
I think this is wrong. IIUC it doesn't replace the internal `epinio-server` prefix but the namespace `epinio` value instead. So in the end the URL is wrong `http://epinio-server.auth.svc.cluster.local/`.

richard-cox replied:
This was an optimisation for the external scenario.

What's the correct internal url (as in, where does the `auth.<domain>` external url point to), something like...?
```
http://dex.epinio.svc.cluster.local
```

thehejik replied:
yes `http://dex.epinio.svc.cluster.local` should be correct but it's not listening on port `80` but on `5556` and `5558`.

Ingress for `auth.<domain>` points to its port `5556`.

richard-cox committed:
Merge remote-tracking branch 'origin/main' into dex-pr

richard-cox committed:
Redirect succesful auth calls directly to the UI rather than Rancher redirect shim
- Rancher API / local dashboard builds redirect `verify-auth` to `auth/verify`
  - In Rancher API this handles redirecting to ember or vue
- Without these the user just gets redirected back to log in page
- For our use case we can just redirect the user to the epinio ui

richard-cox committed:
Improvements
- Fix stripping of query params on redirect to /auth/verify
- Add support to disable Dex
- Customise client secret

enrichman commented on src/jetstream/plugins/epinio/dex/api.go:
```diff
@@ -0,0 +1,43 @@
+		)
+	}
+
+	dexUrl, _ := oidcProvider.AuthCodeURLWithPKCE(state)
```
I have the feeling that you should not use the PKCE flow, since you have a secret and it's a server-to-server request. You can probably use directly the `oidcProvider.Config.AuthCodeURL(state)`

https://dexidp.io/docs/using-dex/

The PKCE flow is needed when you cannot embed a secret safely (in our cli for example, or in a mobile application). If it works you can drop a lot of unneeded files.

richard-cox replied:
Ah, yeah. The secret wasn't that safe before but now it is.

Is there any inherent downside to leaving this in for the moment?

enrichman replied:
The PKCE is a bit less secure I guess. It's not because of it, but just because usually temporary means forever.

Do you think is possible to do a quick change and test? If it's complicated we can try to revisit in a second quick round. We could maybe do this by ourself, and your review.

enrichman replied:
Hum, having a look [here](https://www.scottbrady91.com/oauth/client-authentication-vs-pkce) maybe it's even better to have both. So let's keep it like this!
